### PR TITLE
simulates delay

### PR DESCRIPTION
**Pull request by JustinObanor**

### CODE REVIEW

File: kufar/widget/main.go [hunk 1]
```diff
@@ -25,6 +26,11 @@ func main() {
 		go func(num int) {
 			defer wg.Done()
 			c <- widget{label: "widget_" + strconv.Itoa(num), time: time.Now()}
```
Comment (JustinObanor) [inline kufar/widget/main.go:42]: not sure if i need this line still

File: kufar/widget/main.go [hunk 2]
```diff
@@ -13,32 +14,52 @@ type widget struct {
 	time  time.Time
 }
 
-var n = flag.Int("n", 1, "how many widgets produced by producer")
+var wid = flag.Int("w", 1, "how many widgets produced by producer")
+var con = flag.Int("c", 1, "widgets consumed")
```
Comment (bynov) [inline kufar/widget/main.go]: It's number of consumers :P

File: kufar/widget/main.go [hunk 3]
```diff
@@ -13,32 +14,52 @@ type widget struct {
...
 		close(c)
 	}()
-
-	consumer(c)
 }
 
-func consumer(c chan widget) {
-	for elem := range c {
-		fmt.Printf("[%s  %v]\n", elem.label, elem.time.Format("15: 04: 05.0000"))
+func consumer(ctx context.Context, c chan widget, wg *sync.WaitGroup, con int) {
```
Comment (bynov) [inline kufar/widget/main.go]: You don't need a context here and a select

**Review by bynov — CHANGES_REQUESTED** [review]
You should create two separate wgs. One for producer and second for consumers. When all consumers will exit from loop, you should call `wg.Done()`.
And in main func you should stop your program when `wg.Wait()` for consumer wg's.

File: kufar/widget/main.go [hunk 4]
```diff
@@ -13,32 +13,48 @@ type widget struct {
...
-		}(i)
+			pwg.Done()
+		}(tickCounter)
+
+		tickCounter++
 	}
 
-	go func(){
-		wg.Wait()
+	go func() {
```
Comment (bynov) [inline kufar/widget/main.go:49]: Please remove that go func.

IN your main func it need to be `close(c)` and after that `cwg.Wait` to wait until all consumers will finish with processing

**Review by bynov — APPROVED** [review]

**Pull request merged**